### PR TITLE
feat(path): Isolate Scoop apps' PATH

### LINKED ISSUE #1813: Isolate scoop path to separate environment variable
I have lots of other utilities that are constantly fiddling with my user PATH, and sometimes these utilities and scoop get into fights and wreck things. It'd be nice if scoop could limit all its changes to a single `$env:SCOOP_PATH` variable, which I could interpolate into my own `PATH` at whatever position (and hence precedence) I want.

### PR DESCRIPTION
**Pull request by niheaven**
#### Description

Separate from #5836, will rebase on `develop` when #5836 is merged.

Use the `use_isolated_path` config option to separate Scoop apps PATH to `SCOOP_PATH` (or any other name), then add `%SCOOP_PATH%` to `$env:Path` when installing apps with `env_add_path` for the first time.

#### Motivation and Context

- Closes #1813

- Relates to #5813

This PR does the following:
- Use `-TargetEnvVar` in `Add-Path()` and `Remove-Path()` to add/remove path in different EnvVars
- Check `use_isolated_path` in `env_add_path()` and `env_rm_path()` to add/remove apps path in `%SCOOP_PATH%` or `%PATH%`
- Add `%SCOOP_PATH%` when first add path to it

Path switching is done with `Complete-ConfigChange()`.

#### How Has This Been Tested?

Passed scoop test.
image

`%SCOOP_PATH%` is added correctly:
image
image

#### Checklist:

- [x] I have read the [Contributing Guide](https://github.com/ScoopInstaller/.github/blob/main/.github/CONTRIBUTING.md).
- [ ] I have ensured that I am targeting the `develop` branch.
- [ ] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.
- [x] I have added an entry in the CHANGELOG.

### CODE REVIEW

File: lib/system.ps1 [hunk 1]
```diff
@@ -88,6 +88,7 @@ function Test-PathLikeEnvVar {
 function Add-Path {
     param(
         [string]$Path,
+        [string]$Scope = 'PATH',
```
Comment (chawyehsu) [inline lib/system.ps1:91]: [It's confusing using it for the variable name here](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_scopes). What do you think of `$TargetEnvVar`?

**Comment by chawyehsu** [comment]
Is it a replacement of #5813?

File: lib/system.ps1 (hunk 1 above)
Comment (niheaven) [reply lib/system.ps1:91]: Great, I'll change it.

**Comment by niheaven** [comment]
> Is it a replacement of #5813?

Yes, after reviewing the above PR, I found a simpler and clearer way to implement the feature. This was especially true after refactoring the EnvVar-related functions.

Initially, I attempted to modify pull request #5813, but ultimately decided it would be easier to create a new one instead.

File: lib/install.ps1 [hunk 2]
```diff
@@ -929,13 +935,21 @@ function env_rm_path($manifest, $dir, $global, $arch) {
     $env_add_path = arch_specific 'env_add_path' $manifest $arch
     $dir = $dir.TrimEnd('\')
     if ($env_add_path) {
+        if (get_config USE_ISOLATED_PATH) {
+            $target_path = 'SCOOP_PATH'
+        } else {
+            $target_path = 'PATH'
+        }
```
Comment (chawyehsu) [inline lib/install.ps1:942]: I'm think of a seamless transition of this feature. With current implementation, there are two situations I concern:

- If one turns on `use_isolated_path` and then updates one installed app, in this case the path added to `$env:PATH` by the installed version can't be removed and remains.
- If one turned on `use_isolated_path` and installed one app, then they turn off `use_isolated_path` and try to uninstall the app, in this case the path added to `$env:SCOOP_PATH` can't be removed and remains.

Both situations can be mitigated by checking `$env:PATH` and `$env:SCOOP_PATH` side by side whatever the `use_isolated_path` would be. WDYT?

Comment (niheaven) [reply lib/install.ps1:942]: > For Path switching, I'll make a PR by using https://github.com/ScoopInstaller/Scoop/pull/5851's Complete-ConfigChange().

This is one possible solution, when the user changes the config, all paths will be moved. And of course, removing from all target env sounds reasonable.

Comment (chawyehsu) [reply lib/install.ps1:942]: #5851 is not going to be included in v0.4.0 while this PR will be, so there should be a preliminary solution.

Comment (niheaven) [reply lib/install.ps1:942]: > #5851 is not going to be included in v0.4.0 while this PR will be, so there should be a preliminary solution.

btw, #5851 is ready for review. I've been using it for several days now and everything is working well.

Comment (niheaven) [reply lib/install.ps1:942]: Done. And I'll check if there's a more efficient way.

Comment (niheaven) [reply lib/install.ps1:942]: Aha, okay, I could check all paths in PATH that start with $scoopdir. Will fix it later.

Comment (chawyehsu) [reply lib/install.ps1:942]: Oh you probably misunderstood my point. I didn't mean to force converting all previously added PATH. I meant it should check both `$env:PATH` and `$env:SCOOP_PATH` when doing uninstallation so there won't be paths leftover when users toggle the config back and forth.

Comment (niheaven) [reply lib/install.ps1]: All done. Now switching won't waste any more time and uninstalling will remove paths from `$env:PATH` and `$env:SCOOP_PATH`.

File: lib/core.ps1 [hunk 3]
```diff
@@ -142,11 +142,63 @@ function set_config {
...
+        [string]
+        $Value
+    )
+
+    if ($Name -eq 'use_isolated_path') {
+        . "$PSScriptRoot\..\lib\system.ps1"
+
+        if ($Value -eq $true) {
+            info 'Switching to isolated path... This may take a while, please wait.'
+            $movedPath = Remove-Path -Path "$scoopdir\apps\*" -Quiet -PassThru
```
Comment (chawyehsu) [inline lib/core.ps1:170]: `apps\scoop`?

Comment (niheaven) [reply lib/core.ps1:170]: That dir is not in PATH, or I'm missing sth?

Comment (niheaven) [reply lib/core.ps1:170]: `scoop\shims` is there, and we only shim `scoop.ps1`.

File: lib/install.ps1 [hunk 4]
```diff
@@ -906,19 +906,13 @@ function env_add_path($manifest, $dir, $global, $arch) {
...
-        # GH-3785: Add path in ascending order.
-        [Array]::Reverse($env_add_path)
-        $env_add_path | Where-Object { $_ } | ForEach-Object {
-            if ($_ -eq '.') {
-                $path_dir = $dir
-            } else {
-                $path_dir = Join-Path $dir $_
-            }
-            if (!(is_in_dir $dir $path_dir)) {
-                abort "Error in manifest: env_add_path '$_' is outside the app directory."
```
Comment (chawyehsu) [inline lib/install.ps1:918]: This error report is useful but removed, is it expected?

File: lib/install.ps1 [hunk 5]
```diff
@@ -906,19 +906,13 @@ function env_add_path($manifest, $dir, $global, $arch) {
...
-            }
-            Add-Path -Path $path_dir -Global:$global -Force
+        if (get_config USE_ISOLATED_PATH) {
+            Add-Path -Path '%SCOOP_PATH%' -Global:$global
+            $target_path = 'SCOOP_PATH'
+        } else {
+            $target_path = 'PATH'
         }
+        Add-Path -Path ($env_add_path.Where({ $_ -and !$_.Contains(':') }).ForEach({ Join-Path $dir $_ })) -TargetEnvVar $target_path -Global:$global -Force
```
Comment (chawyehsu) [inline lib/install.ps1:915]: What's the reason of filtering entry having `:`?

File: lib/install.ps1 [hunk 6]
```diff
@@ -906,19 +906,13 @@ function env_add_path($manifest, $dir, $global, $arch) {
...
-            if (!(is_in_dir $dir $path_dir)) {
-                abort "Error in manifest: env_add_path '$_' is outside the app directory."
-            }
-            Add-Path -Path $path_dir -Global:$global -Force
+        if (get_config USE_ISOLATED_PATH) {
+            Add-Path -Path '%SCOOP_PATH%' -Global:$global
+            $target_path = 'SCOOP_PATH'
+        } else {
+            $target_path = 'PATH'
         }
```
Comment (chawyehsu) [inline lib/install.ps1:914]:
```suggestion
        $target_path = 'PATH'
        if (get_config USE_ISOLATED_PATH) {
            Add-Path -Path '%SCOOP_PATH%' -Global:$global
            $target_path = 'SCOOP_PATH'
        }
```

File: lib/install.ps1 (hunk 4 above)
Comment (niheaven) [reply lib/install.ps1:918]: It's clear that `$path_dir` is `Join-Path $dir $_` and is in `$dir`, so the test here could never be run.

File: lib/install.ps1 (hunk 6 above)
Comment (niheaven) [reply lib/install.ps1:914]: Hmm, is there any difference between these two implements?

Comment (chawyehsu) [reply lib/install.ps1:914]: hoisting variable, less loc

File: lib/install.ps1 (hunk 5 above)
Comment (chawyehsu) [reply lib/install.ps1:915]: Then what about relative path like `../<path>` that escapes from it? Before this it was checked by `is_in_dir` but you removed it in https://github.com/ScoopInstaller/Scoop/pull/5840/files#r1559002809

Comment (niheaven) [reply lib/install.ps1:914]: Wow, you're right. I'll add it back.

Comment (niheaven) [reply lib/install.ps1]: Fix it now and add some tests.

**Comment by rashil2000** [comment]
I think we should use isolated path just by default (instead of making it a config option), seeing as it doesn't seem to have any disadvantages. Or does it, maybe I missed something?

**Comment by CEbbinghaus** [comment]
This limits the path to SCOOP_PATH correct? Wouldn't it be better to allow the user to choose which name the environment variable takes?

**Comment by niheaven** [comment]
> This limits the path to SCOOP_PATH correct? Wouldn't it be better to allow the user to choose which name the environment variable takes?

For convenience, a fixed name is better for users.

> I think we should use isolated path just by default (instead of making it a config option), seeing as it doesn't seem to have any disadvantages. Or does it, maybe I missed something?

Yes, it is, and we still need a migration. @ScoopInstaller/maintainers How about this feature?

**Comment by rashil2000** [comment]
> For convenience, a fixed name is better for users.

I agree, it's sufficient and self-explanatory.

> Yes, it is, and we still need a migration. @ScoopInstaller/maintainers How about this feature?

Migration will be needed. We can even update the Readme and the website intro, saying that PATH pollution is now handled better because SCOOP_PATH is modified now, instead of directly modifying PATH.

**Comment by chawyehsu** [comment]
> I think we should use isolated path just by default (instead of making it a config option)

We need a transition period.

**Comment by rashil2000** [comment]
What for, if I may ask? 
The `scoop update` command can handle the environment change without user intervention, right? It doesn't seem to be a breaking change.

**Comment by niheaven** [comment]
By setting `SCOOP_PATH` as default, I will make a commit and do migration as `Complete-ConfigChange()`.

File: lib/install.ps1 [hunk 7]
```diff
@@ -906,34 +906,23 @@ function env_add_path($manifest, $dir, $global, $arch) {
...
-                $path_dir = $dir
-            } else {
-                $path_dir = Join-Path $dir $_
-            }
-            if (!(is_in_dir $dir $path_dir)) {
-                abort "Error in manifest: env_add_path '$_' is outside the app directory."
-            }
-            Add-Path -Path $path_dir -Global:$global -Force
+        $target_path = 'PATH'
+        if (get_config USE_ISOLATED_PATH) {
```
Comment (rashil2000) [inline lib/install.ps1:910]: This config is still there - intentional?

Comment (niheaven) [reply lib/install.ps1:910]: By mistake 🤣

**Comment by niheaven** [comment]
Got the idea. So it's better to offer config now and notify users that we'll change the default behavior and change it a few months later.

What's your opinion? @chawyehsu @rashil2000

**Comment by rashil2000** [comment]
I don't have an issue with providing a config option per se - I just don't see purpose in involving user interaction in an internal implementation detail.

**Comment by niheaven** [comment]
So choose an implementation here and then we can start bumping the version.

If config, I'll revert the last commit and wait for several months to submit a PR again, if default, then directly merge all commits.

File: lib/core.ps1 [hunk 8]
```diff
@@ -142,11 +142,28 @@ function set_config {
...
+    param (
+        [Parameter(Mandatory, Position = 0)]
+        [string]
+        $Name,
+        [Parameter(Mandatory, Position = 1)]
+        [AllowEmptyString()]
+        [string]
+        $Value
+    )
+    # Add needed config change logic here
```
Comment (rashil2000) [inline lib/core.ps1:164]: An empty function?

File: libexec/scoop-update.ps1 [hunk 9]
```diff
@@ -57,6 +57,22 @@ if(($PSVersionTable.PSVersion.Major) -lt 5) {
 }
 $show_update_log = get_config SHOW_UPDATE_LOG $true
 
+# Migrate PATH EnvVar to SCOOP_PATH
+$movedPath = Remove-Path -Path "$scoopdir\apps\*" -Quiet -PassThru
```
Comment (rashil2000) [inline libexec/scoop-update.ps1:61]: Will this be run every time on `scoop update`?

File: lib/core.ps1 (hunk 8 above)
Comment (niheaven) [reply lib/core.ps1:164]: Yes, empty function, for later usage.

File: libexec/scoop-update.ps1 (hunk 9 above)
Comment (niheaven) [reply libexec/scoop-update.ps1:61]: Yes, check every time since we don't know if the user has mixed PATH. But could be removed in next version.

Comment (rashil2000) [reply libexec/scoop-update.ps1:61]: I see, makes sense.

File: lib/core.ps1 (hunk 8 above)
Comment (rashil2000) [reply lib/core.ps1:164]: Umm, so we'll add it when we have usage? Right now it doesn't fit here.

Comment (niheaven) [reply lib/core.ps1:164]: Hmm, see usage in #5851. will remove from here now

**Comment by rashil2000** [comment]
So, I'm facing a weird error: 
image

Did you face/rectify this?

**Comment by rashil2000** [comment]
In addition to `$scoopdir\apps\*` paths, the SCOOP_PATH should also contain `$scoopdir\shims`, right? I think we missed it here...

**Comment by niheaven** [comment]
No, putting `$scoopdir\shims` directly in PATH should be preferred. This will not break scoop's feature in case of `SCOOP_PATH`-related error.

**Comment by CEbbinghaus** [comment]
> > For convenience, a fixed name is better for users.
>
> I agree, it's sufficient and self-explanatory.

I am not arguing that it is not self explanatory. I am saying that limiting it to 1 single environment variable breaks peoples environments who use that environment variable for other purposes (e.g setting the path to the scoop executable). It is fairly easy to allow configuration of the environment variable so its not a huge cost while the benefit to end users might be not having to fix their setup.

**Comment by chawyehsu** [comment]
>use that environment variable for other purposes (e.g setting the path to the scoop executable)

`SCOOP_PATHS`/`SCOOP_APPPATHS` would be a better name if you concern about this, I don't think `SCOOP_PATHS`/`SCOOP_APPPATHS` has been used elsewhere. It is feasible to allow variable env var with something like `use_isolated_path: "MY_PERFECT_ENV_NAME"` but I'm not a fan of this kind of flexibility.

> What's your opinion?
>
> So choose an implementation here and then we can start bumping the version.
>
> If config, I'll revert the last commit and wait for several months to submit a PR again, if default, then directly merge all commits.

Abstention

**Review by chawyehsu — APPROVED** [review]

**Pull request merged**